### DISCUSSION TITLE
Gem5 code style formatting with .clang-format

### DISCUSSION THREAD

**Discussion by vspetrov** [gem5/gem5#333] [gem5-dev]
Hello,
my collegues and I have been working on Gem5 project. We want to adjust to upstream(community) coding style of Gem5 (described here https://www.gem5.org/documentation/general_docs/development/coding_style/). However, it looks like post-commit hook is not the most convenient way to adjust to code style. For example, having set of formatting rules summarized in .clang-format would allow very convenient and automatic integration with external tools, e.g. VSCode.

So, the question: is there any plans in community to work out set of code styling rules via .clang-format ? Would community be interested in that (if .clang-format is made as close as possible to the documented style) ?

**Comment by giactra** [comment]
Good question. Some people tried in the past (starting from @powerjg ) to define a clang-format file (see: https://gem5-review.googlesource.com/c/public/gem5/+/49063). Last attempt was from Tom Rollet. You can gather from the gerrit discussion what was their blocker.

IMHO any PR with a clang-format file (ideally based on the one posted for review on gerrit) would be much appreciated.

**Comment by powerjg** [comment]
Yes, this would be a huge contribution!!!

My prior attempt to use clang-format got derailed because it wasn't perfect. We shouldn't let perfect be the enemy of progress, though!

I think the biggest change that needs to be made to my post is that there were a number of places we should turn off clang-format (e.g., when defining `BitSet`s). I tried to make clang-format understand this automatically, but wasn't able to get it to work, and I don't think it was worth the effort. We should just manually disable it in those cases.

There's also useful information in my email to gem5-dev. I'll copy paste below. https://www.mail-archive.com/gem5-dev@gem5.org/msg40017.html

------

For fun(?), I have come up with a clang-format file that's pretty darn
close to our current style.

However, the problem is that most of our files don't actually meet our
style guide! :'( So, running this clang-format generates an enormous
changeset. Skimming through the (100,000+) changes clang-format makes all
of them look "correct" to me. But I didn't check them all, just a few
spot-checks.

I've put together a WIP commit as an example of what it looks like to run
clang-format on gem5. There are a few caveats, however.

1. BitUnion really confuses clang-format for two reasons: (1) We indent the
code inside the BitUntion macro. (2) there are no semicolons after the
macro calls for BitUnion and EndBitUntion.
2. There are a few lines that go over 79 characters, but they're easy to
manually fix (it's only 4 files).
3. When I ran style.py on all files in src/ there were a huge number that
had invalid sorting of includes. In at least one case, this causes compiler
failures when I let style.py fix it.
4. I'm skipping sorting includes with clang-format right now. There is a
way to specify the blocks that we define, but I don't want to dive into
this right now. I'll create a Jira issue so we can come back to it.

Finally, I'm using the following shell script to test things. This runs sed
to tell clang-format to ignore the BitUnion blocks, runs clang-format, and
then undoes the sed script.

If this seems like something we want to continue with, then I'll commit
these changes and we can add clang-format to the commit hooks.

Clang-format file:
https://gem5-review.googlesource.com/c/public/gem5/+/49063
The giant changeset:
https://gem5-review.googlesource.com/c/public/gem5/+/49064/

The script I used (don't judge my sed ability :D):
```
# Make clang-format ignore bitunions
grep -r -l BitUnion src | xargs sed -i -e '/^ *BitUnion.*(.*)/i \
// clang-format off' -e '/^ *EndBitUnion(.*)/a; \
// clang-format on'

# undo the change to the bitunion file
git checkout src/base/bitunion.hh

# Run clang format
find src -name "*.cc" -or -name "*.hh" -exec clang-format -style=file -i {}
\;

# Remove clang-format off in BitUnion files
grep -r -l BitUnion src | xargs sed -i -e '/^ *\/\/ clang-format.*$/d' -e
'/^; *$/d'
```

**Reply by AntonTimofeev** [reply]
clang-format for some reason works fine with BitUnion macros, when these macros are removed from StatementMacros list

**Comment by BobbyRBruce** [comment]
I'm very much in favor of `clang-format` being used. I'd actually like `clang-format` to be our style guide for C/CPP. That is if our `clang-format` setup does not complain about something then it should be seen as being in-accordance with our style guide. This removes any human judgement from the proces. When we started using [Python Black](https://pypi.org/project/black/) to automatically format and check our Python Code we decided that Black was our style guide for Python. If Black is ok with something, it is fine, if not Black corrects it. I've found this simplifies things for us and would like the same for C/CPP (note: we still have to manually check variable names, class names, and some string lengths but this isn't so bad... would be nice if we could automate it too though).
 
So with this, I'm ok with updating our style guide in accordance to what we can do with `clang-format` rather than trying too hard to replicate our current style guide as much as possible. My question which goes one step further is **would it perhaps be beneficial to take an "off-the-shelf" `clang-format` configuration?** I notice [here](https://clang.llvm.org/docs/ClangFormatStyleOptions.html) that `clang-format` comes with some preset styles: LLVM, Google, Chromium, Mozilla, WebKit, Microsoft. If memory serves me correctly gem5's style guide is based on Google's. I'd vote for doing this to simplify things. 

Annoyingly there's no standard C/CPP style-guide and there's no shortage of people who can spend hours debating the finer points of line-lengths and `int& x` over `int &x`. I'm of the opinion that getting a decent, mature style checker in gem5 is more important than having a style guide that's perfect, and taking something "off-the-shelf" at least side-steps this discussion a bit. It's also nice to say, directly, what style guide we are using rather than explain our own home-brewed approach to new contributors.

**Comment by andysan** [comment]
While it would be really nice to use an off-the-shelf configuration like Google's style guide, I think there is a clear risk that it would create far too much trouble to be worth it. I'm particularly concerned about merge issues when backporting changes and rebasing old changes. Not to mention various non-public branches with niche models.

I would suggest that we compare the diffs from the off-the-shelf styles that are closest to the current style to Jason's configuration (which seemed very close). If the difference is small, we go with the off-the-shelf configuration. If not, we use Jason's config (we might want to add the missing include block specifications) and update the style guide to define that style as the golden reference.

**Comment by vspetrov** [comment]
thanks for the discussion everybody. Looks like there at least a concensus that making style available via clang-format is a way to go. On our side, we will evaluate Jason's config as well and return with feedback.

**Comment by abmerop** [comment]
Sort of echoing what @andysan said. In addition I am curious what would happen for files that are currently not following the style guide.  The easiest example here is Ruby.  If we have clang-format and I make a Ruby change is it going to try to force me to re-style the code around it?  Maybe this isn't an issue but I am not familiar with clang-format.

**Reply by powerjg** [reply]
I believe the original proposal was to bite the bullet and run clang-format on everything in one commit.

**Reply by AntonTimofeev** [reply]
I've got a crazy idea to use https://github.com/newren/git-filter-repo/blob/main/contrib/filter-repo-demos/lint-history and re-format all commits from the beginning. I tried this and it actually worked and took ~20 minutes on my laptop. But includes sorting has to be turned off because some header files are not including what they use.

**Reply by BobbyRBruce** [reply]
That's a nice idea but it involves changing literally every commit. At the very lease it'd mean if you `git pull`'d your local repo after this change git would be confused as to where all these altered commits  came from cause all sorts of merge issues. I imagine this would break a lot of other things for people too. Imagine, if you will, a script which tries to checkout gem5 at a particular revision. With this, that revision no longer exists. Generally in gem5, with git, we rebase/edit our commit history locally, merge it into `develop` and never alter it again. It causes too many problems to do anything else.

**Reply by BobbyRBruce** [reply]
>Sort of echoing what @andysan said. In addition I am curious what would happen for files that are currently not following the style guide. The easiest example here is Ruby. If we have clang-format and I make a Ruby change is it going to try to force me to re-style the code around it? Maybe this isn't an issue but I am not familiar with clang-format.

I don't think we're going with changing the style-guide but if we were, as Jason mentioned, we'd run across the whole codebase first. The main argument against this is it ruins `git blame`. There is a "[.git-blame-ignore-revs](https://github.com/gem5/gem5/blob/48a40cf2f5182a82de360b7efa497d82e06b1631/.git-blame-ignore-revs)" file which can fix this (basically it says to ignore certain commits when running `git blame`). However, I've found using this file is not fully standard across all IDEs and environments. In particular, recognizing this file and using it is only supported "in beta" by GitHub: https://github.blog/changelog/2022-03-24-ignore-commits-in-the-blame-view-beta/.

I'd love to cleanup our style guide (using something "off-the-shelf") but if the new style guide is too dramatic and alters too much code, I'll pass on it for now.

**Comment by trollet91** [comment]
Hello,
I pushed a draft MR for clang format with what I've done last year. #362 
I don't have bandwith to work on this ATM, anyone can take the lead on this MR and change my commits if needed :)

**Comment by mshiryaev** [comment]
@BobbyRBruce , @powerjg ,
As far as I understand from this thread and https://github.com/gem5/gem5/pull/362 the idea with enabling of clang-format is not refused but at least there are concerns about backporting/rebasing/non-public branches. Also seems this activity is not high priority so there is limited bandwidth for this.

Could you please clarify when you expect clang-format can be enabled in gem5?
Any rough estimation would be helpful (like in few months or in specific release).

**Reply by powerjg** [reply]
We haven't discussed any specific timeline. @BobbyRBruce Do you want to try to get this done for v23.1. Or, alternatively, we could do it as the first commit of develop-for-v24.0

**Reply by BobbyRBruce** [reply]
It'd be really nice to get this in for v23.1. If we wait for after the release we the `stable` and `develop` branches may have  big divergence due to `develop` having been reformatted with clang-format and `stable` not. To do this we'd need this change in my December 1st.

((That being said, it would be managable and doing a code-base reformat as a minor gem5 release would be fine.))

So, yes, I'll try to help get this moving forward. We're so close with it so hopefully we can get it done in the next week.

**Comment by wkjung** [comment]
If you care about the back porting issues and merge hell, how about just adding `git-clang-format` pre-commit hook instead of `clang-format`? This checks ONLY the added/staged lines and leave the other untouched lines, so it does not change other than `.pre-commit-config.yaml` at first and not requires massive changes in all files. And only the following PRs' modifications, not the entire files, will be checked. Not to mention that it, as in clang format, fixes up automatically so a contributor wont suffer any pain from manually fixing formats.

For example https://github.com/tgross35/mirrors-clang-format/blob/844e1d349f6e2c3b2dbb4ec7e80697dcb590589c/.pre-commit-hooks.yaml#L12

**Reply by abmerop** [reply]
I was not aware of this possibility, thanks for point it out!  This option would potentially save me a lot of time to do other things other than fixing merge conflicts.  Will wait for thoughts from other maintainers though.

**Reply by powerjg** [reply]
I don't want to dismiss this idea. It's a good idea.

My worry is that we'll end up with files that have a mix of styles if we do this. Though, I guess that's not too much worse that what we have today where files don't a totally consistent style.

**Reply by pranith** [reply]
Yes, please! Any improvement in this direction is better than the current state.

**Comment by rjc-arch** [comment]
I have prototyped the `git-clang-format` idea mentioned above, and submitted prototypes of two implementations as PRs #2313 and #2314.

It seems to work very well, and has the same behaviour as the Python `black` integration (re-formats the staged C/C++ files, and leaves the modifications unstaged for separate review).

To answer @powerjg's comment above, personally I think that, if we treat this as an aid to ease the code review process, then it would still be beneficial even though we may end up with a slight mix of styles. Without the need to re-format all the gem5 source files in one go, the choice of `clang-format` configuration becomes an easier process, and we can tweak the configuration incrementally over time.
